// egg-ts-helper/dist/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const package_json_1 = tslib_1.__importDefault(require("../package.json"));
exports.tmpDir = path_1.default.resolve(__dirname, '../.tmp');
exports.eggInfoPath = path_1.default.resolve(exports.tmpDir, './eggInfo.json');
exports.dtsCommentRE = new RegExp(`^\\/\\/ [\\w ]+ ${package_json_1.default.name}(@\\d+\\.\\d+\\.\\d+)?`);
exports.dtsComment = `// This file is created by ${package_json_1.default.name}@${package_json_1.default.version}\n` +
    '// Do not modify this file!!!!!!!!!\n';
// mapping declaration in egg
exports.declMapping = {
    service: 'IService',
    controller: 'IController',
    ctx: 'Context',
    context: 'Context',
    app: 'Application',
    application: 'Application',
    agent: 'Agent',
    request: 'Request',
    response: 'Response',
    helper: 'IHelper',
    middleware: 'IMiddleware',
    config: 'EggAppConfig',
    plugin: 'EggPlugin',
};
